apps/kg/train_pytorch.py

```diff
@@ -147,3 +214,128 @@ def train_mp(args, model, train_sampler, valid_samplers=None, rank=0, rel_parts=
 @thread_wrapped_func
 def test_mp(args, model, test_samplers, rank=0, mode='Test', queue=None):
     test(args, model, test_samplers, rank, mode, queue)
+
+@thread_wrapped_func
+def dist_train_test(args, model, train_sampler, entity_pb, relation_pb, l2g, rank=0, rel_parts=None, cross_rels=None, barrier=None):
+    if args.num_proc > 1:
+        th.set_num_threads(args.num_thread)
+
+    client = connect_to_kvstore(args, entity_pb, relation_pb, l2g)
+    client.barrier()
+    train_time_start = time.time()
+    train(args, model, train_sampler, None, rank, rel_parts, cross_rels, barrier, client)
+    print('Total train time {:.3f} seconds'.format(time.time() - train_time_start))
+    client.barrier()
+
+    model = None
+
+    if client.get_id() % args.num_client == 0: # pull full model from kvstore
+
+        args.num_test_proc = args.num_client
+        dataset_full = get_dataset(args.data_path, args.dataset, args.format)
+
+        print('Full data n_entities: ' + str(dataset_full.n_entities))
+        print("Full data n_relations: " + str(dataset_full.n_relations))
+
+        model_test = load_model(None, args, dataset_full.n_entities, dataset_full.n_relations)
+        eval_dataset = EvalDataset(dataset_full, args)
+
+        if args.test:
+            model_test.share_memory()
+
+        if args.neg_sample_size_test < 0:
+            args.neg_sample_size_test = dataset_full.n_entities
+        args.eval_filter = not args.no_eval_filter
+        if args.neg_deg_sample_eval:
+            assert not args.eval_filter, "if negative sampling based on degree, we can't filter positive edges."
+
+        if args.neg_chunk_size_valid < 0:
+            args.neg_chunk_size_valid = args.neg_sample_size_valid
+        if args.neg_chunk_size_test < 0:
+            args.neg_chunk_size_test = args.neg_sample_size_test
+
+        print("Pull relation_emb ...")
+        relation_id = F.arange(0, model_test.n_relations)
+        relation_data = client.pull(name='relation_emb', id_tensor=relation_id)
+        model_test.relation_emb.emb[relation_id] = relation_data
+ 
+        print("Pull entity_emb ... ")
+        # split model into 100 small parts
+        start = 0
+        percent = 0
+        entity_id = F.arange(0, model_test.n_entities)
+        count = int(model_test.n_entities / 100)
+        end = start + count
+        while True:
+            print("Pull %d / 100 ..." % percent)
```

Comment:
use tqdm?

Reply:
That's a good idea. But I'm afraid that it will bring a new tqdm package installed by pip.